HPCC-20149 ESP support for Dynamic ESDL Proxy mode

afishbeck opened the pull request:
Dynamic ESDL should support a mixture of locally configured service
methods and methods that are proxied off to other ESP instances.

This will make the migration of existing environments much easier
to manage.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeck commented:
@wangkx @mayx please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b7ce967d5ae2014b57fcad846c80ef9ca3b25f8c
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |790 |790 |0 |
|test (thor) |721 |721 |0 |
|test (roxie) |861 |861 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   14 sec (00:00:14) |  174 sec (00:02:54) |   61 sec (00:01:01) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1436 sec (00:23:56) |   18 sec (00:00:18) | 1742 sec (00:29:02) |

wangkx commented on esp/bindings/http/client/httpclient.cpp:
```diff
@@ -405,13 +405,134 @@ int CHttpClient::sendRequest(const char* method, const char* contenttype, String
+            StringArray pair;
+            pair.appendList(copyFrom.queryHeaders().item(i), ":", true);
+            const char *name = pair.item(0);
+            const char *value = pair.item(1);
```
Do you need the 'value' here?

afishbeck replied:
Good catch I use pair.item(1) below so can remove it.

wangkx commented on esp/bindings/http/client/httpclient.cpp:
```diff
@@ -405,13 +405,134 @@ int CHttpClient::sendRequest(const char* method, const char* contenttype, String
+
+    if (m_readTimeoutSecs)
+        httpresponse->setTimeOut(m_readTimeoutSecs);
+    httpresponse->receive(false, me);  // MORE - pass in IMultiException if we want to see exceptions (which are not fatal)
```
Do we need to 'forward' the me into forwardResponse?

afishbeck replied:
Good point. Hmmm I'll have to look into what these represent.  At least make sure they are logged if not forwarded.

wangkx commented on esp/services/esdl_svc_engine/esdl_binding.cpp:
```diff
@@ -1812,7 +1932,31 @@ int EsdlBindingImpl::onGet(CHttpRequest* request, CHttpResponse* response)
+                    nextPathNode(thepath, method);
+                    StringBuffer proxyAddress;
+                    if (checkForMethodProxy(root, method, proxyAddress))
+                        return forwardProxyMessage(proxyAddress, request, response);
```
'break;'?

afishbeck replied:
yes, should add it.

wangkx reviewed (commented):
@afishbeck some minor comments/questions.

mayx reviewed (commented):
Don't see any issue with the coding. A couple of general comments:
1) In the Proxy section, maybe the @address can be renamed to "@forwardTo" or something like that? It can be confusing as to whether the address is being used a proxy, vs actually the current esp is acting as a proxy to forward to the @address specified. 
2) How can we prevent someone misconfiguring the @address to point back to itself, thus creating a dire situation of endless loop? Maybe can utilize x-forwarded-for or some kind of custom header?

mayx commented:
Maybe can utilize "x-forward-for" or a custom header?

afishbeck commented:
@mayx agreed will change @ address.  And I should look into adding x-forwarded-for in any case.  Including perhaps to prevent endless loop.

afishbeck commented:
@afishbeck keep ownership

afishbeck committed:
HPCC-20149 ESP support for Dynamic ESDL Proxy mode

Dynamic ESDL should support a mixture of locally configured service
methods and methods that are proxied off to other ESP instances.

This will make the migration of existing environments much easier
to manage.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

afishbeck commented:
@wangkx @mayx updated based on your past comments.  Please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: fa6f65d6ea6dc3829eaff6c661384dcca78b3ad9
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  182 sec (00:03:02) |   93 sec (00:01:33) |   21 sec (00:00:21) |   43 sec (00:00:43) | 1327 sec (00:22:07) |   19 sec (00:00:19) | 1694 sec (00:28:14) |

afishbeck commented:
@wangkx @mayx please do final review.

wangkx commented:
Looks fine for me.

mayx reviewed (approved):
Looks good.
